include/SH3/system/glvertarray.hpp

```diff
@@ -0,0 +1,118 @@
+ *
+ *  Also gives us a nice abstract and typesafe way to deal with VAOs.
+ *
+ *  @copyright 2017 Palm Studios
+ *
+ *  @date 8-6-2017
+ *
+ *  @author Jesse Buhagiar
+ */
+#ifndef GLVERTARRAY_HPP_INCLUDED
+#define GLVERTARRAY_HPP_INCLUDED
+
+#include <vector>
+
+#include "GL/glew.h"
+#include "GL/gl.h"
+
+#include "SH3/system/glbuffer.hpp"
+
+namespace sh3_gl
+{
+    /**
+     *  A wrapper for an OpenGL 3.3 Vertex Array Object.
+     */
+    struct sh3_glvertarray final
+    {
+    public:
+
+        sh3_glvertarray(){Create();};
+
+        /**
+         *  VAO Attribute locations. These are used in a @ref sh3_glprogram to access data
+         *  sent to it from a VBO. These values were picked for no specific reason other than
+         *  the usual order of a vertex (vert data, normal data, uv data).
+         */
+        enum Attribute
+        {
+            SH3_VERTEX_SLOT = 0,
+            SH3_NORMAL_SLOT = 1,
+            SH3_UV_SLOT = 2,
+            SH3_INDEX_SLOT = 3
+        };
+
+        /**
+         *  Specifies data type of data passed to paramater 3 of glVertexAttribPointer.
+         */
+        enum DataType
+        {
+            SH3_BYTE = GL_BYTE,
+            SH3_UBYTE = GL_UNSIGNED_BYTE,
+            SH3_SHORT = GL_SHORT,
+            SH3_USHORT = GL_UNSIGNED_SHORT,
+            SH3_INT = GL_INT,
+            SH3_UINT = GL_UNSIGNED_INT,
+            SH3_FLOAT = GL_FLOAT
+        };
+
+        /**
+         *  Create a Vertex Attribute Object
+         */
+        void Create();
+
+        /**
+         *  Realease our VAO.
+         */
+        void Release();
+
+        /**
+         *  Bind this VAO.
+         */
+        void Bind();
+
+         /**
+          * Unind this VAO.
+          */
+        void Unbind();
+
+          /**
+           * Get the ID of this VAO.
+           */
+        GLuint GetID() const {return id;}
+
+        /**
+         *  Add a buffer to our VAO.
+         */
+        void AddVBO(sh3_glbuffer& buff);
+
+        /**
+         *  Set vertex shader location of data in VBO for specified attribute.
+         *
+         *  @param attrib - Attribute we want to map data input in our shader (using location = x).
+         *  @param type - Data Type of data passed in.
+         *  @param size - The size of each component per attribute.
+         *  @param stride - Byte offset between vertex attributes.
+         *  @param offset - Offset to attribute in buffer. Each attrib is then located at offset + stride.
+         */
+        void SetDataLocation(Attribute attrib, DataType type, GLint size, GLsizei stride, GLint offset);
+
+        /**
+         *  Draw our VertexArray
+         */
+        void Draw();
+
+
+    private:
+        GLuint                      id; /**< This VAO's ID. */
+        std::vector<sh3_glbuffer>   buffers; /**< All VBOs managed by this VAO. MUST BE IN ORDER SPECIFIED BY @ref Attribute. */
```

Comment:
`sh3_glbuffer` seems small enough that we could just store an `std::array` of sh3_glbuffers with the maximum possible number of buffers, which is also a reasonably small 4 (this constant could also be included in the `Attribute` `enum` as `MAX` or something.
We then need an additional member to keep track of which buffers have been initialized. I guess an `std::underlying_type_t<Attribute>` which gets incremented when calling AddVBO does the job.